### PR TITLE
Fix 9449

### PR DESCRIPTION
**Pull request by dsyme**
While integrating #6810 into #6811 a regression turned up.  

The problem lies in the fact that an extra constraint is being asserted during optimization.  The relevant lines of #6810 are things like [this](https://github.com/dotnet/fsharp/pull/6810/files#diff-641da7c52ffc888699888ebd03559cb5R1948) and [this].  The tril fix removes these lines which were not needed.  Extra constraints should *not* need to be asserted and [this](https://github.com/dotnet/fsharp/pull/6810/files#diff-641da7c52ffc888699888ebd03559cb5R3084) addition later in the development of #6810 meant these other two changes are not needed.

TODO:

* [x] add testing around the regression

Note the regression is active even if `langversion` is not set, which also indicates some inaccurate coding in #6810

### CODE REVIEW

**Comment by dsyme** [comment]
@KevinRansom @cartermp This is now ready (once green).  Since this is a regression it's worth expediting.

**Comment by dsyme** [comment]
Unfortunately CI did not test this fix properly - the "green tick" ✔️ was a false one as no tests ran.  I'll have to look at this more closely now

**Comment by TIHan** [comment]
There is one test that is failing at the moment.

**Comment by dsyme** [comment]
> There is one test that is failing at the moment.

RIght, that's because the fix didn't work :-)))

I've pushed a different fix (the same code is removed, but the problem that code was addressing is now dealt with differently). 

Notes:

* The "CodegenWitness..." code is a late rediscovery of the solutions to constraints

* To do this for a call to a method `M<T1, ... TN>` at type arguments `M<ty1...tyN>` we freshen type parameters `T1...TN` and assert equalities `T1 = ty1` ... `TN = tyN`, propagating constraints as we asset each equality, and see what constraint solutions were found

* The process of asserting these constraints one by one causes problems where half way through there appear to be unmet constraints.  Previously I'd worked around this incorrectly by allowing these unmet constraints to be added to the type parameters during codegen. This just isn't right.

* Instead, we should assert the equalities simultaneously *then* propagate constraints. 

Basically when you have a complicated tangle of constraints involving a bunch of type parameters there needs to be simultaneous processing here
